### ISSUE TITLE
support animation

### ISSUE THREAD

**Issue by jkroso** [jkroso/mana#2]
Probably it should look something like this:

```js
const from = <div style={{left:0}}/>
const to = <div style={{left:5}}/>
<Animation duration='100ms' from={from} to={to}/>.mount(someDivElement)
```

or more likely:

```js
from.mount(someDivElement).update(<Animation duration='100ms'>{to}</Animation>)
```

and if the animation node has its `update` method called before it finishes animating it should attempt to adopt the new node as it's `to`

**Comment by gaydenko** [comment]
To my understanding, animation is a part of more wide area, that is asynchronization :) At the moment `Element` termination looks like:
```js
  remove() {
    notifyUnmount(this)
    super.remove()
  }
```
With heavy asynchronous apps, notification isn't sufficient. To make any asynchronous operation possible (not only animation related ones) there is an old common idiom (pseudo-code):
```js
element.onRemove = done => {
    doSomeCleanupWork();
    done();
}
```
That is real removing is fulfilling by supplied parameter `done`.

Just after `remove()` implementation there is `replace()` method - and it is an example of possible async use case: `next` can not be mounted before cleanup work is done by "current".

Well, it's my opinion. And what do you think?

**Comment by jkroso** [comment]
Animation is not conceptually an async thing in the context of a functional UI lib. Even though under the hood it would be using async things. The conceptual model I want you to have when thinking of animations is that you are declaring a sequence of UI nodes up front. Each node in this sequence represents what the UI looks like after a certain amount of time has passed. (1ms, 2ms, etc...)

With Mana your use case of delaying the display of the next node until the current node has successfully cleaned up after itself should be implemented by manipulating the state and letting the changes in state cause the change in UI.
 
```js
state.isDoneCleaningUp
  ? <span>done!</span>
  : <Spinner/>
```

Animations can't really use state because an animation is a living thing; it has a predetermined future and any snapshot you take during the animation will not capture this future. And I want my state to be snapshot-able because that enables saving state, syncing it across users/machines, and undo/redo. 

As I think about this more and more I realise it would actually be possible to do animation robustly via state but it would be difficult and hard to optimise. For now a special `Animation` type is fairly easy to make fast and it won't require any special treatment through the rest of Mana so I think I'll still give it a shot.

**Comment by gaydenko** [comment]
Yes, agree, animation is predictable as far as CSS API demands ms constants. And yes, it is noticeable architecture decision - choosing between providing an animation as a particular case of common asynchronous opportunities, or with the simplest direct way. 

I guess, if you have plans to support common async use cases, then direct animation implementation can.. well, complicate further async implementation. So, it can be easier to implement async, and even then to implement animation helpers as just a particular case of async.

At the moment I use another virutal DOM implementation (Jiesa) having async support for node life cycle. And I use it for animation also - it is easy. Say, a helper connecting fade effect with node life cycle takes just 25 lines of code (sorry, can not publish) dealing mostly with style itself.
